OrbitLinuxTracing/ContextSwitchAndThreadStateVisitor.cpp

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "ContextSwitchAndThreadStateVisitor.h"
+
+#include <absl/container/flat_hash_map.h>
+
+#include <algorithm>
+#include <utility>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitLinuxTracing/TracerListener.h"
+
+namespace LinuxTracing {
+
+using orbit_grpc_protos::SchedulingSlice;
+using orbit_grpc_protos::ThreadStateSlice;
+
+void ContextSwitchAndThreadStateVisitor::ProcessInitialTidToPidAssociation(pid_t tid, pid_t pid) {
+  bool new_insertion = tid_to_pid_association_.insert_or_assign(tid, pid).second;
+  if (!new_insertion) {
+    ERROR("Overwriting previous pid for tid %d with initial pid %d", tid, pid);
+  }
+}
+
+void ContextSwitchAndThreadStateVisitor::visit(ForkPerfEvent* event) {
+  pid_t pid = event->GetPid();
+  pid_t tid = event->GetTid();
+  bool new_insertion = tid_to_pid_association_.insert_or_assign(tid, pid).second;
+  if (!new_insertion) {
+    ERROR("Overwriting previous pid for tid %d with pid %d from PERF_RECORD_FORK", tid, pid);
```

Comment:
That's why it's not a `CHECK`, but in the time of a capture this is definitely strange enough to be an `ERROR`. It can also happen if the fork is recorded while still reading from `/proc`, which is related to the flakiness of the other comment but still worthy to be noted as `ERROR`.